lib/ansible/executor/task_executor.py

```diff
@@ -605,8 +606,10 @@ def _execute(self, variables=None):
 
         if module_context is not None:
             module_defaults_fqcn = module_context.resolved_fqcn
+        elif self._task.action is not None:
+            module_defaults_fqcn = self._task.action
```

Comment:
This is not correct, the task action is whatever the user passed in, not the resolved name.

Reply:
why not just https://github.com/ansible/ansible/commit/266483cd2d56b13e381c3e88b16cfbc01546a5a0#diff-37f95accdc58a04484ee270a7a95c0652c9774ae9657e398323b7909c4215901R612-R615? I don't know what this is trying to fix.

Reply:
task.resolved_action was only needed for normal.py. If the action loads the defaults for the module implicitly like you suggested, all TE needs to do is load the defaults for the handler (excluding handlers designated by a module's `action_plugin` runtime.yml metadata, since it's a toggle to avoid this behavior):

If it's implicitly done for action plugin module execution (rather just normal.py), this could be simplified to:
```diff
-        if module_context is not None:
-            module_defaults_fqcn = module_context.resolved_fqcn
-        else:
-            module_defaults_fqcn = self._task.resolved_action
-
-        # Apply default params for action/module, if present
-        self._task.args = get_action_args_with_defaults(
-            module_defaults_fqcn, self._task.args, self._task.module_defaults, templar,
-            action_groups=self._task._parent._play._action_groups
-        )
+        if module_context is None:
+            # Apply default params for action, if present
+            self._task.args = get_action_args_with_defaults(
+                self._handler.ansible_name, self._task.args, self._task.module_defaults, templar,
+                action_groups=self._task._parent._play._action_groups
+            )
```

Otherwise if it's opt in for action plugins, and normal.py opts in:

```diff
         if module_context is not None:
             module_defaults_fqcn = module_context.resolved_fqcn
         else:
-            module_defaults_fqcn = self._task.resolved_action
+            module_defaults_fqcn = self._handler.ansible_name
 
```

If we want to change resolved_action for callback plugins to follow the same logic as module_defaults (which is my question here https://github.com/ansible/ansible/pull/82003#issuecomment-1767012158), then it could used here too (but I'm still partial to the action plugin consolidation).